bot/utils/redis_dict.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+import json
+from collections.abc import MutableMapping
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import redis as redis_py
+
+from bot import constants
+
+ValidRedisKey = Union[str, int, float]
+JSONSerializableType = Optional[Union[str, float, bool, Dict, List, Tuple, Enum]]
+
+
+class RedisDict(MutableMapping):
+    """
+    A dictionary interface for a Redis database.
+
+    Objects created by this class should mostly behave like a normal dictionary,
+    but will store all the data in our Redis database for persistence between restarts.
+
+    Redis is limited to simple types, so to allow you to store collections like lists
+    and dictionaries, we JSON deserialize every value. That means that it will not be possible
+    to store complex objects, only stuff like strings, numbers, and collections of strings and numbers.
+    """
+
+    _namespaces = []
+    _redis = redis_py.Redis(
+        host=constants.Bot.redis_host,
+        port=constants.Bot.redis_port,
+    )  # Can be overridden for testing
+
+    def __init__(self, namespace: Optional[str] = None) -> None:
+        """Initialize the RedisDict with the right namespace."""
+        super().__init__()
+        self._has_custom_namespace = namespace is not None
+
+        if self._has_custom_namespace:
+            self._set_namespace(namespace)
+        else:
+            self.namespace = "global"
```

Comment:
Is this supposed to set `self._namespace`?

Reply:
oh. yeah it is. good eye! will fix.